File: AddressBook/react-client/src/dtos/State.ts
import Common from "../utility/Common";
import { CityRes } from "./City";
import { CountryRes } from "./Country";
import { PagedReq } from "./PagedReq";

export interface StateRes {
  stateId?: string;
  name?: string;
  code?: string;
  latitude?: number;
  longitude?: number;

  countryId?: string;
  country?: CountryRes;

  cities?: CityRes[];
}

export class StateReqEdit {
  name?: string = "";
  code?: string = "";
  latitude?: number = 0;
  longitude?: number = 0;

  countryId?: string = "";
  constructor(countryId?: string) {
    this.countryId = countryId;
  }
}

export class StateReqSearch extends PagedReq {
  countryId?: string;
  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy = "",
      searchText = "",
    }: PagedReq,
    {countryId = ""}
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.countryId = countryId;
  }
}

File: AddressBook/Dtos/Dtos/State.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using AddressBook.Common.Paging;

namespace AddressBook.Dtos
{
    public class StateRes
    {
        public int StateId { get; set; }
        public string? Name { get; set; }
        public string? Code { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }

        // Foreign keys
        public int? CountryId { get; set; }
        public CountryRes? Country { get; set; }

        // Child tables
        public ICollection<CityRes>? Cities { get; set; }
    }

    public class StateReqEdit
    {
        [Required]
        public string? Name { get; set; }
        [Required]
        public string? Code { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }

        // Foreign keys
        public int? CountryId { get; set; }
    }

    public class StateReqSearch : PagedReq
    {
        public int? CountryId { get; set; }
    }
}
